compositor_pipeline/src/pipeline/output/whip/payloader.rs

```diff
@@ -0,0 +1,322 @@
+use bytes::Bytes;
+use std::{collections::VecDeque, fmt::Debug};
+use tracing::error;
+use webrtc_util::Marshal;
+
+use rand::Rng;
+use rtp::codecs::{h264::H264Payloader, opus::OpusPayloader};
+
+use crate::pipeline::{
+    rtp::{AUDIO_PAYLOAD_TYPE, VIDEO_PAYLOAD_TYPE},
+    types::{EncodedChunk, EncodedChunkKind},
+    AudioCodec, VideoCodec,
+};
+
+const H264_CLOCK_RATE: u32 = 90000;
+const OPUS_CLOCK_RATE: u32 = 48000;
+
+struct RtpStreamContext {
+    ssrc: u32,
+    next_sequence_number: u16,
+    received_eos: bool,
+}
+
+impl RtpStreamContext {
+    pub fn new() -> Self {
+        let mut rng = rand::thread_rng();
+        let ssrc = rng.gen::<u32>();
+        let next_sequence_number = rng.gen::<u16>();
+
+        RtpStreamContext {
+            ssrc,
+            next_sequence_number,
+            received_eos: false,
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum PayloadingError {
+    #[error("Tried to payload video with non video payloader.")]
+    NoVideoPayloader,
+
+    #[error("Tried to payload audio with non audio payloader.")]
+    NoAudioPayloader,
+
+    #[error(
+        "Tried to payload video with codec {:#?} with payloader for codec {:#?}",
+        chunk_codec,
+        payloader_codec
+    )]
+    NonMatchingVideoCodecs {
+        chunk_codec: VideoCodec,
+        payloader_codec: VideoCodec,
+    },
+
+    #[error(
+        "Tried to payload audio with codec {:#?} with payloader for codec {:#?}",
+        chunk_codec,
+        payloader_codec
+    )]
+    NonMatchingAudioCodecs {
+        chunk_codec: AudioCodec,
+        payloader_codec: AudioCodec,
+    },
+
+    #[error(transparent)]
+    RtpLibError(#[from] rtp::Error),
+
+    #[error(transparent)]
+    MarshalError(#[from] webrtc_util::Error),
+
+    #[error("Audio EOS already sent.")]
+    AudioEOSAlreadySent,
+
+    #[error("Video EOS already sent.")]
+    VideoEOSAlreadySent,
+}
+
+pub struct Payloader {
+    video: Option<VideoPayloader>,
+    audio: Option<AudioPayloader>,
+}
+
+enum VideoPayloader {
+    H264 {
+        payloader: H264Payloader,
+        context: RtpStreamContext,
+    },
+}
+
+enum AudioPayloader {
+    Opus {
+        payloader: OpusPayloader,
+        context: RtpStreamContext,
+    },
+}
+
+pub enum DataKind {
+    Audio,
+    Video,
+}
+
+pub struct Payload {
+    pub data: Bytes,
+    pub kind: DataKind,
+}
+
```

Comment:
Those types seem unnecessary, if anything I would just add

```suggestion
pub struct Payload {
    Video(Bytes),
    Audio(Bytes),
}
```